pkg/driver/provisioner.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2021 The Ceph-COSI Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+You may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package driver
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	radosgwapi "github.com/ceph/go-ceph/rgw/admin"
+	"github.com/thotz/cosi-driver-ceph/pkg/util/s3client"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"k8s.io/klog/v2"
+	cosi "sigs.k8s.io/container-object-storage-interface-spec"
+)
+
+// contains two clients
+// 1.) for RGWAdminOps : mainly for user related operations
+// 2.) for S3 operations : mainly for bucket related operations
+type ProvisionerServer struct {
+	provisioner        string
+	S3Client           *s3client.S3Agent
+	radosgwAdminClient *radosgwapi.API
+}
+
+// ProvisionerCreateBucket is an idempotent method for creating buckets
+// It is expected to create the same bucket given a bucketName and protocol
+// If the bucket already exists, then it MUST return codes.AlreadyExists
+// Return values
+//    nil -                   Bucket successfully created
+//    codes.AlreadyExists -   Bucket already exists. No more retries
+//    non-nil err -           Internal error                                [requeue'd with exponential backoff]
+func (s *ProvisionerServer) ProvisionerCreateBucket(ctx context.Context,
+	req *cosi.ProvisionerCreateBucketRequest) (*cosi.ProvisionerCreateBucketResponse, error) {
+	klog.Info("Using ceph rgw to create Backend Bucket")
+	protocol := req.GetProtocol()
+	if protocol == nil {
+		klog.ErrorS(fmt.Errorf("Invalid Argument"), "Protocol is nil")
+		return nil, status.Error(codes.InvalidArgument, "Protocol is nil")
+	}
+	s3 := protocol.GetS3()
+	if s3 == nil {
+		klog.ErrorS(fmt.Errorf("Invalid Argument"), "S3 protocol is nil")
+		return nil, status.Error(codes.InvalidArgument, "S3 Protocol is nil")
+	}
+	//TODO : validate S3 protocol defined, check points valid rgwendpoint, v4 signature check etc
+	bucketName := req.GetName()
+	klog.V(3).InfoS("Create Bucket", "name", bucketName)
+
+	err := s.S3Client.CreateBucket(bucketName)
+	if err != nil {
+		// Check to see if the bucket already exists by above api
+		klog.ErrorS(err, "Bucket creation failed")
```

Comment:
```suggestion
		klog.ErrorS(err, "failed to create bucket")
```

Also pass the `bucketName`?